### PR TITLE
raftstore: optimize write priority scheduling path

### PR DESCRIPTION
**Pull request by Connor1996**
### What is changed and how it works?

Issue Number: Ref https://github.com/tikv/tikv/issues/14353

```commit-message
Optimize write priority scheduling path including:
- replace `DashMap` with `RwLock` in the resource controller 
- use visit pattern for consuming msg resource to avoid constructing hashmap
- introduce `ParsedEntry` to avoid parsing raft command from entry data repeatedly in different places
```

### Check List 

Tests 

- Unit test
- Integration test

Have 1% improvement with resource control enabled, compared to master. 
image

### Release note 

```release-note
optimize write priority scheduling path performance overhead
```

### CODE REVIEW

File: components/raftstore/src/store/util.rs [hunk 1]
```diff
@@ -745,6 +751,139 @@ pub(crate) fn u64_to_timespec(u: u64) -> Timespec {
     Timespec::new(sec as i64, nsec as i32)
 }
 
+// ParsedEntry wraps raft-proto `Entry` and used to avoid parsing raft command
+// from entry's data repeatedly. The parsed command may be used in multiple
+// places, so cache it at the first place.
+pub struct ParsedEntry {
```
Comment (nolouch) [inline components/raftstore/src/store/util.rs:757]: consider a better name? maybe as use `Entry` , `RaftEntry` or `LogEntry` directly?

Comment (Connor1996) [reply components/raftstore/src/store/util.rs:757]: These names can't be distinguished with the current `raftpb::Entry`

**Review by glorv — APPROVED** [review]
LGTM

**Review by nolouch — APPROVED** [review]
lgtm

File: components/raftstore/src/store/util.rs (hunk 1 above)
Comment (nolouch) [reply components/raftstore/src/store/util.rs:757]: It's a little strange with parsed.

Comment (Connor1996) [reply components/raftstore/src/store/util.rs:757]: maybe call it `LazyEntry`?

Comment (Connor1996) [reply components/raftstore/src/store/util.rs:757]: `Parsed` describes the behavior, that it provides parsed items (cmd and conf_change) to outers. `parse` is just an inner function.

Comment (nolouch) [reply components/raftstore/src/store/util.rs:757]: What about `PlainEntry` or `DecodedEntry`?

Comment (Connor1996) [reply components/raftstore/src/store/util.rs:757]: `Entry` is defined in proto, I can't change the name.

**Review by tabokie — APPROVED** [review]

**Pull request merged**